exporter/exporterhelper/persistent_storage.go

```diff
@@ -0,0 +1,442 @@
+	}
+}
+
+// loop is the main loop that handles fetching items from the persistent buffer
+func (pcs *persistentContiguousStorage) loop() {
+	// We want to run it here so it's not blocking
+	reqs := pcs.retrieveUnprocessedItems(context.Background())
+	if len(reqs) > 0 {
+		errCount := 0
+		for _, req := range reqs {
+			if pcs.put(req) != nil {
+				errCount++
+			}
+		}
+		pcs.logger.Info("moved items for processing back to queue",
+			zap.String(zapQueueNameKey, pcs.queueName),
+			zap.Int(zapNumberOfItems, len(reqs)), zap.Int(zapErrorCount, errCount))
+	}
+
+	for {
+		for {
+			req, found := pcs.getNextItem(context.Background())
+			if found {
+				select {
+				case <-pcs.stopChan:
+					return
+				case pcs.reqChan <- req:
+				}
+			} else {
+				select {
+				case <-pcs.stopChan:
+					return
+				case <-time.After(pcs.retryDelay):
+				}
+			}
+		}
+	}
+}
+
+// get returns the request channel that all the requests will be send on
+func (pcs *persistentContiguousStorage) get() <-chan request {
+	return pcs.reqChan
+}
+
+// size returns the number of currently available items, which were not picked by consumers yet
+func (pcs *persistentContiguousStorage) size() int {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+	return int(pcs.writeIndex - pcs.readIndex)
+}
+
+// numberOfCurrentlyProcessedItems returns the count of batches for which processing started but hasn't finish yet
+func (pcs *persistentContiguousStorage) numberOfCurrentlyProcessedItems() int {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+	return len(pcs.currentlyProcessedItems)
+}
+
+func (pcs *persistentContiguousStorage) stop() {
+	pcs.logger.Debug("stopping persistentContiguousStorage", zap.String(zapQueueNameKey, pcs.queueName))
+	pcs.stopOnce.Do(func() {
+		_ = currentlyProcessedBatchesGauge.UpsertEntry(func() int64 {
+			return int64(0)
+		}, metricdata.NewLabelValue(pcs.queueName))
+		close(pcs.stopChan)
+	})
+}
+
+// put marshals the request and puts it into the persistent queue
+func (pcs *persistentContiguousStorage) put(req request) error {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	ctx := context.Background()
+
+	itemKey := pcs.itemKey(pcs.writeIndex)
+	if !pcs._clientSet(ctx, itemKey, req, requestToBytes) {
+		return errStoringItemToQueue
+	}
+
+	pcs.writeIndex++
+	if !pcs._clientSet(ctx, writeIndexKey, pcs.writeIndex, itemIndexToBytes) {
+		return errUpdatingIndex
+	}
+
+	return nil
+}
+
+// getNextItem pulls the next available item from the persistent storage; if none is found, returns (nil, false)
+func (pcs *persistentContiguousStorage) getNextItem(ctx context.Context) (request, bool) {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	if pcs.readIndex < pcs.writeIndex {
+		index := pcs.readIndex
+		// Increase here, so even if errors happen below, it always iterates
+		pcs.readIndex++
+
+		pcs.updateReadIndex(ctx)
+		pcs.itemProcessingStart(ctx, index)
+
+		req := pcs._clientGet(ctx, pcs.itemKey(index), pcs.bytesToRequest).(request)
```

Comment:
And perhaps even this `Get` can be batched with the 2 previous `Set` calls (not sure if that provides any performance benefits though, worth checking).
If there are such benefits then we are not just looking for a simple batched `Set` but rather a transaction of Sets and Gets (similar to what bolt interface supports).
What do you think?